frontend/catalyst/utils/wrapper.cpp

```diff
@@ -1,15 +1,186 @@
+
+    size_t memref_len = ranks.attr("__len__")().cast<size_t>();
+    size_t offset = 0;
+
+    char *memref_array_bytes = (char *)(memref_array_ptr);
+
+    for (size_t idx = 0; idx < memref_len; idx++) {
+        unsigned int rank_i = ranks.attr("__getitem__")(idx).cast<unsigned int>();
+        char *memref_i_beginning = memref_array_bytes + offset;
+        offset += memref_size_based_on_rank(rank_i);
+
+        struct memref_beginning_t *memref = (struct memref_beginning_t *)memref_i_beginning;
+        bool is_in_rt_heap = f_transfer_ptr(memref->allocated);
+
+        if (!is_in_rt_heap) {
+            // This case is guaranteed by the compiler to be the following:
+            // when an input tensor is sent to as an output as well.
+            //
+            // It is guaranteed by the use of the flag --cp-global-memref
+            //
+            // Use the numpy_arrays dictionary which sets up the following map:
+            // integer (memory address) -> py::object (numpy array)
+            auto array_object = numpy_arrays.attr("__getitem__")((size_t)memref->allocated);
+            returns.append(array_object);
+            continue;
+        }
+
+        const npy_intp *dims = npy_get_dimensions(memref_i_beginning, rank_i);
+
+        size_t element_size = sizes.attr("__getitem__")(idx).cast<size_t>();
+        const npy_intp *strides = npy_get_strides(memref_i_beginning, element_size, rank_i);
+
+        auto etype_i = etypes.attr("__getitem__")(idx);
+        PyArray_Descr *descr = PyArray_DescrFromTypeObject(etype_i.ptr());
+        if (!descr) {
+            throw std::runtime_error("PyArray_Descr failed.");
+        }
+
+        void *aligned = memref->aligned;
+        PyObject *new_array =
+            PyArray_NewFromDescr(&PyArray_Type, descr, rank_i, dims, strides, aligned, 0, NULL);
+        if (!new_array) {
+            throw std::runtime_error("PyArray_NewFromDescr failed.");
+        }
+
+        PyObject *capsule =
+            PyCapsule_New(memref->allocated, NULL, (PyCapsule_Destructor)&free_wrap);
+        if (!capsule) {
+            throw std::runtime_error("PyCapsule_New failed.");
+        }
+
+        int retval = PyArray_SetBaseObject((PyArrayObject *)new_array, capsule);
+        bool success = 0 == retval;
+        if (!success) {
+            throw std::runtime_error("PyArray_SetBaseObject failed.");
+        }
+
+        returns.append(new_array);
+
+        // Now we insert the array into the dictionary.
+        // This is to take into account the possibility
+        // of returning more than 1 array with the same buffer.
+        // Although, it looks like it won't happen in reality.
+        PyObject *pyLong = PyLong_FromLong((size_t)memref->allocated);
+        if (!pyLong) {
+            throw std::runtime_error("PyLong_FromLong failed.");
+        }
+
+        numpy_arrays[pyLong] = new_array;
+
+        Py_DECREF(pyLong);
+        Py_DECREF(new_array);
```

Comment:
Just confirming something here, but why are we decrementing the new_array reference count here?
Is it because on creation the count is 1, then inserting it into the capsule bumps it to 2, and so we need to reduce to 1 before exiting (that way after the capsule is destroyed its count will be 0 and it will be destroyed as well)?

Reply:
Close. We decrement the reference because `new_array` exits the scope. The capsule does not increment the reference. What increments the reference is adding it to the list (1 + 1 = 2) and adding it to the dictionary (2 + 1 = 3). At the end we want to keep the references for the list and the dictionary but we still need to decrement one for the local scope.

Reply:
Thanks

Reply:
Yes. I ran scalene and saw the necessary improvements. Will run again and show the output here.

Reply:
For a state of 256M:

![Screenshot from 2023-05-10 12-40-36](https://github.com/PennyLaneAI/catalyst/assets/110487834/5c9fae34-37de-4768-8346-ded9b4d719ab)

No memory leaks reported by scalene.

Reply:
Does it show where the returned buffer is deallocated (or at least reference counted to 0)?

Reply:
I don't believe that is a functionality available in scalene. But if it were not deallocated / the reference counter reached zero, we would see bar graph go up. To provide more context, the example I used is the following:

```python
@qjit()
@qml.qnode(qml.device("lightning.qubit", wires=24))
def f():
    return qml.state()


for i in range(20):
    f()
```

Another indicator (which I didn't showed previously) is the total memory used. Here it is:

![Screenshot from 2023-05-10 13-21-25](https://github.com/PennyLaneAI/catalyst/assets/110487834/354adbee-87c3-4773-bc0b-7bf191965077)

I also ran the program after removing the `Py_DECREF(new_array)` line and we see this instead. We can compare both images:

![Screenshot from 2023-05-10 13-15-17](https://github.com/PennyLaneAI/catalyst/assets/110487834/e90baff6-0c40-4c1d-95e2-8e8a1c45b4a9)

![Screenshot from 2023-05-10 13-18-47](https://github.com/PennyLaneAI/catalyst/assets/110487834/14692f6f-2a6b-4337-b646-b7da28fb0c67)

Therefore the `Py_DECREF` correctly decrements the array. If we were to add one more `Py_DECREF(new_array)` line, the program would crash.

Reply:
For completion, the other ~256MB come from a buffer allocated during the device initialization. We never use this buffer.

![Screenshot from 2023-05-10 13-26-00](https://github.com/PennyLaneAI/catalyst/assets/110487834/7ed5cd26-597b-479f-baa8-79226a0934b3)

Reply:
Excellent, thank you @erick-xanadu !